bpf/bpf_lxc.c

```diff
@@ -1602,18 +1602,31 @@ ipv4_policy(struct __ctx_buff *ctx, int ifindex, __u32 src_label, enum ct_status
 #endif /* ENABLE_PER_PACKET_LB && !DISABLE_LOOPBACK_LB */
 
 #ifdef ENABLE_NODEPORT
-	if (ret == CT_NEW || ret == CT_REOPENED) {
+	{
 		bool dsr = false;
 # ifdef ENABLE_DSR
 		int ret2;
 
-		ret2 = handle_dsr_v4(ctx, &dsr);
+		ret2 = handle_dsr_v4(ctx, &dsr, (enum ct_status)ret);
 		if (ret2 != 0)
 			return ret2;
 
 		ct_state_new.dsr = dsr;
 		if (ret == CT_REOPENED && ct_state.dsr != dsr)
 			ct_update4_dsr(get_ct_map4(&tuple), &tuple, dsr);
+
+#if DSR_ENCAP_MODE == DSR_ENCAP_IPIP_CNI
+		if (!revalidate_data(ctx, &data, &data_end, &ip4))
+			return DROP_INVALID;
+		tuple.nexthdr = ip4->protocol;
+		tuple.daddr = ip4->daddr;
+		tuple.saddr = ip4->saddr;
+
+		ret = ct_lookup4(get_ct_map4(&tuple), &tuple, ctx, l4_off, CT_INGRESS, &ct_state,
+				 &monitor);
```

Comment:
Why do we need to lookup the connection state again here when we already performed a lookup on line 1529?

Reply:
The tuple changed. The lookup at line 1529 is looking for outer tuples. After handle_dsr_v4 on line 1610, the nexthdr and saddr changed. If I don't do lookup here, it still works. But there will be duplicate ct entries for a single connection.